### PR TITLE
Extend PalletInfoAccess with module_name and crate_version method

### PR DESCRIPTION
**Pull request by KiChjang**
Per request from @gavofyork via Element, this PR adds module names and crate versions to the `PalletInfoAccess` struct as public fields.

### CODE REVIEW

File: frame/support/src/traits/metadata.rs
Comment (bkchr) [inline frame/support/src/traits/metadata.rs:54]: IMO this doesn't make any sense here.

`CallMetadata` is about an instance of a call. Now you adding `crate_version` and the `index` which only works if this is an outer call.

`index` and `crate_version` should be moved above to `PalletInfoAccess`.

Comment (KiChjang) [reply frame/support/src/traits/metadata.rs:54]: That's what I thought too, I wasn't sure where it would make the most sense, but I'll try putting it in `PalletInfoAccess`. Thanks!

Comment (bkchr) [reply frame/support/src/traits/metadata.rs:54]: Ty! :)

**Review by thiolliere — APPROVED** [review]
Code is good to me (top message should be updated to reflect the current state of the PR)

I don't know if this will fix Gav request. Because contrary to method CallMetadata which list all the modules, this PR only allow to get the crate for a specific module, it is not possible to list all modules index and crate versions as far as I can see.

(But we should be able to do so by introducing a new trait similar to `CallMetadata`. The easiest would be to implement this trait on Tuple of type which implements PalletInfoAccess, and then we can simply use the type alias `AllPalletsWithSystem`. (or we can implement manually like `CallMetadata`))

**Comment by gavofyork** [comment]
Yeah this doesn't actually do what I need (yet). But the PR description does describe what I need. @KiChjang do you intend to get this to the point where it'a in line with the PR description?

**Comment by bkchr** [comment]
@gavofyork do you have an actual instance of a call on which you want to get this metadata?

**Comment by gavofyork** [comment]
@bkchr No - I only have a string which is meant to be the pallet's name. With this string I need to be able to get all pallet instances of that name together with their index and crate version.

Since a pallet instance has two associated names - the module/crate name (e.g. `"pallet_balances"`, `"crowdloan"`) and the instance name (e.g. `"Balances"`, `"Crowdloan"`), ideally I would get both.

**Comment by bkchr** [comment]
Yeah, so `CallMetadata` wouldn't work here anyway.

@KiChjang the only thing that is missing is the module/crate_name which could be added to `PalletInfo` as well. Then we should be able in Polkadot to create a trait that will be implemented for tuples of `PalletInfo` types aka `AllPalletsWithSystem` and then we should be able to collect all the data that @gavofyork needs for XCM.

File: frame/support/src/lib.rs
Comment (bkchr) [inline frame/support/src/lib.rs:672]:
```suggestion
/// This means that the returned string will correspond to the name of the crate the macro is
```

**Review by bkchr — COMMENTED** [review]
For the name the idea was more to take the input to `construct_runtime`. Because stuff like `crowdloan`, `claims` etc all come from the same crate.

File: frame/support/src/traits/metadata.rs
Comment (thiolliere) [inline frame/support/src/traits/metadata.rs:33]: Maybe we should update the doc of PalletInfo.

PalletInfo was introduced as: 
```
/// Provides information about the pallet setup in the runtime.
///
/// An implementor should be able to provide information about each pallet that
/// is configured in `construct_runtime!`.
```

It was meant to associate information given to construct_runtime for each pallets. The index and the name of the pallet are defined by construct_runtime, that is why there was a need to introduce this trait and the associated type in frame_system::Config::PalletInfo.

And PalletInfoAccess, was implemented to easily access the information in frame_system::Config::PalletInfo.

With this PR there is a bit of interdependency between both traits: PalletInfoAccess will call PalletInfo for name and index, and PalletInfo will call PalletInfoAccess for crate_name and crate_version.

I think it is fine, but the doc of PalletInfo and PalletInfoAccess should be updated to something like:
```
/// Provides information about the pallet itself and its setup in the runtime.
///
/// An implementor should be able to provide information about each pallet that
/// is configured in `construct_runtime!`.
pub trait PalletInfo {
```
```
/// Provides information about the pallet itself and its setup in the runtime.
///
/// Declare some information and access the information provided by [`PalletInfo`] for a specific
/// pallet.
pub trait PalletInfoAccess {
```

I'm ok to have this difficult relation as long as both traits are meant to be implemented by pallet macro and construct_runtime

Comment (bkchr) [reply frame/support/src/traits/metadata.rs:33]: Good point. I think version should be moved to `PalletInfoAccess` and `PalletInfo` gives the name/module_path as I said above.

**Comment by KiChjang** [comment]
> For the name the idea was more to take the input to `construct_runtime`. Because stuff like `crowdloan`, `claims` etc all come from the same crate.

The issue here is that whatever was provided to `construct_runtime!`, i.e. this identifier here...

```rust
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: system::{Pallet, Call, Event<T>, Origin<T>} = 30,
		NestedModule3: nested::module3::{Pallet, Call, Config, Storage, Event, Origin},
		               ^^^^^^^^^^^^^^^
	}
```

... is not necessarily a single identifier, and so grabbing the crate name directly from the pallet is the easiest way to disambiguate. I suppose I could also make a naming scheme where the double colons get converted to underscores, so we'd have `nested_module3` as the identifier for such a crate.

File: frame/support/test/tests/construct_runtime.rs
Comment (KiChjang) [inline frame/support/test/tests/construct_runtime.rs:976]: This one here is a bit odd. I have half a mind to instruct the module_name method to ignore `self` and `crate` keywords when it sees them appearing at the beginning of the path...

**Comment by gavofyork** [comment]
> I suppose I could also make a naming scheme where the double colons get converted to underscores, so we'd have nested_module3 as the identifier for such a crate.

If it's just going into a string, then you can keep the `::`s.

**Comment by gavofyork** [comment]
It's not such a huge issue that naming is a bit idiosyncratic when dealing with pallets that are one of several in a crate local to the project. The primary use-case for this is for standardised pallets anyway, so if they have necessarily non-standard paths then they're unlikely to be the target of this code.

**Comment by thiolliere** [comment]
I think the pallet path given to construct_runtime shouldn't really identify anything other than the path to access this pallet. People can do renaming, or reexport etc...

Instead if xcm want to check if a pallet is made from some specific pallet "blueprint" like council is made from the "blueprint" pallet-collective. Maybe we need to allow the pallet to declare what is his "blueprint". The crate name as said above is not precise enough, and it is a bit weird to me to rely on the path which should just give some regular rust path to some crate.

Instead pallet could be able to declare their "blueprint" name maybe ? we can do that with an optional new attribute `#[pallet::blueprint = "foo"]` and so we would have for each pallet an `Option<Blueprint>`.

File: frame/support/test/tests/construct_runtime.rs
Comment (thiolliere) [reply frame/support/test/tests/construct_runtime.rs]: I'm not sure, what if the path is exaclty `crate` ?

**Comment by gavofyork** [comment]
The primary use-case is to identify pallet instances from standard pallet types; I don't think adding yet another identifier is going to help with that. As long as there's a crate name and crate version then we should be good.

**Comment by KiChjang** [comment]
This needs another approval before it can be merged, anything else I need to add here?

File: frame/support/src/traits/metadata.rs
Comment (thiolliere) [inline frame/support/src/traits/metadata.rs:32]: maybe we should precise which rust module name. because a rust module is not unique
```suggestion
	/// Convert the given pallet `P` into its Rust module name as used in the runtime construction.
```

**Review by thiolliere — APPROVED** [review]
looks good to me, identifying a pallet kind using its rust module may be fine. Note that this doesn't include the specification of the instance, so 2 pallets can have the same module name.

**Comment by gavofyork** [comment]
Needs resolving and CI failing @KiChjang

**Review by gavofyork — APPROVED** [review]

**Comment by gavofyork** [comment]
A test would have been nice.

**Pull request merged**
